File: server/posts.py
from flask import Blueprint, jsonify, request, send_file
import os

import web.queries.posts as queries
from db.services.posts.actions import (
    toggle_selected as _toggle_selected,
    delete_post as _delete_post,
)

posts_bp = Blueprint("posts", __name__, url_prefix="/api/posts")


def _nav_context(row_id, all_rows):
    idx = next((i for i, (rid, _) in enumerate(all_rows) if rid == row_id), None)
    if idx is None:
        return None
    total = len(all_rows)
    return {
        "current": idx + 1,
        "total": total,
        "prev_id": all_rows[idx - 1][0] if idx > 0 else None,
        "next_id": all_rows[idx + 1][0] if idx < total - 1 else None,
    }


@posts_bp.get("/next")
def next_unprocessed():
    row_id = queries.get_first_unprocessed_id()
    if row_id is None:
        return jsonify({"row_id": None})
    return jsonify({"row_id": row_id})


@posts_bp.get("/<int:row_id>")
def get_post(row_id):
    post = queries.get_post_by_id(row_id)
    if not post:
        return jsonify({"error": "not found or already processed"}), 404

    post = dict(post)
    post["unprocessed_count"] = queries.count_unprocessed_by_author(post["author"])

    all_rows = queries.get_unprocessed_ids()
    nav = _nav_context(row_id, all_rows)

    return jsonify({"post": post, "nav": nav})


@posts_bp.get("/author/<int:row_id>")
def get_author_posts(row_id):
    author = queries.get_author_of_post(row_id)
    if author is None:
        return jsonify({"error": "not found"}), 404

    all_posts = queries.get_posts_by_author_unprocessed(author)
    if not all_posts:
        return jsonify({"author": author, "post": None, "nav": None})

    target = dict(next((p for p in all_posts if p["id"] == row_id), all_posts[0]))
    target["unprocessed_count"] = len(all_posts)

    all_rows = [(p["id"], p["processed"]) for p in all_posts]
    nav = _nav_context(target["id"], all_rows)

    return jsonify({"author": author, "post": target, "nav": nav})


@posts_bp.post("/<int:row_id>/mark")
def mark(row_id):
    processed = int(request.json["processed"])
    queries.update_processed(row_id, processed)
    return jsonify({"ok": True})


@posts_bp.post("/<int:row_id>/selected")
def selected(row_id):
    value = int(request.json["selected"])
    _toggle_selected(row_id, value)
    return jsonify({"ok": True})


@posts_bp.delete("/<int:row_id>")
def delete(row_id):
    return jsonify({"ok": _delete_post(row_id)})

File: frontend/src/api/posts.ts
// frontend/src/api/posts.ts
import { createRequest } from "./http";

const request = createRequest("/api/posts");

export interface Post {
  id: number;
  post_id: string;
  author: string;
  text: string;
  post_url: string;
  processed: number;
  group_name: string;
  scraped_at: string;
  selected: number;
  result_json_v1: string | null;
  gate1_reasoning: string | null;
  gate1_prompt_version: string | null;
  extraction_result_json: string | null;
  extraction_reasoning: string | null;
  extraction_prompt_version: string | null;
  unprocessed_count: number;
}

export interface NavContext {
  current: number;
  total: number;
  prev_id: number | null;
  next_id: number | null;
}

export interface PostWithNav {
  post: Post;
  nav: NavContext;
}

export interface AuthorPosts {
  author: string;
  post: Post | null;
  nav: NavContext | null;
}

export function getNextUnprocessed(): Promise<{ row_id: number | null }> {
  return request("/next");
}

export function getPost(rowId: number): Promise<PostWithNav> {
  return request(`/${rowId}`);
}

export function getAuthorPosts(rowId: number): Promise<AuthorPosts> {
  return request(`/author/${rowId}`);
}

export function markProcessed(
  rowId: number,
  processed: 0 | 1,
): Promise<{ ok: boolean }> {
  return request(`/${rowId}/mark`, {
    method: "POST",
    body: JSON.stringify({ processed }),
  });
}

export function toggleSelected(
  rowId: number,
  selected: 0 | 1,
): Promise<{ ok: boolean }> {
  return request(`/${rowId}/selected`, {
    method: "POST",
    body: JSON.stringify({ selected }),
  });
}

export function deletePost(rowId: number): Promise<{ ok: boolean }> {
  return request(`/${rowId}`, { method: "DELETE" });
}
